File: ex.pl
#! /usr/bin/env perl

# Fuzzy files explorer

use v5.12;
use warnings;
use utf8;
use re '/aa';
use Getopt::Long qw/GetOptions :config no_ignore_case bundling/;
use Term::ANSIColor qw/color :constants/;
use File::Basename 'fileparse';

# Help
my $help = << '────';
ex [options] [topic]

--(no-)hidden,   -H: include hidden files (default)
--directory=dir, -d: change root directory
--exact,         -e: exact filename matches
--grep,          -g: grep for occurrences of topic in files
--only,          -o: output filetred lines only
--view,          -v: view with your $EDITOR, use alt-Enter from within fzf
────

my ($dir, $hidden) = ('.', 1);

# Options
GetOptions (
   'H|hidden!'     => \   $hidden,
   'd|directory=s' => \   $dir,
   'e|exact'       => \my $exact,
   'g|grep:s'      => \my $grep,
   'o|only'        => \my $only,
   'v|view'        => \my $view,
   'h|help'        => sub { print $help; exit }
) or die RED.'Error in command line arguments'.RESET, "\n";

$dir = glob $dir if $dir =~ /^~/;
$dir =~ s'/+$'';

chdir $dir or die RED.$!.RESET, "\n";

$hidden = $hidden ? '--hidden' : '';

# Functions
sub fzf
{
   my ($query, $grep) = @_;

   $query =~ s/'/'"'"'/g if defined $query; # 'escape' 's

   my $fzf_opts = '-0 -1 --cycle --print-query --expect=alt-enter';
   my @results;

   if (defined $grep and (length $grep or length $query))
   {
      my $pattern;

      if (length $grep) {
         $pattern = $grep;
      } elsif (length $query) {
         $pattern = $query;
      }

      my $find = "rg -FS $hidden -g'!.git' -g'!.svn' -g'!.hg' --ignore-file ~/.gitignore -l '$pattern'";
      my $preview = "rg -FS --color=always '$pattern' {}";

      $preview =~ s/[\\"`\$]/\\$&/g; # quote sh ""s special characters: \ " ` $

      my ($mode, $patt) = ('', '');

      if (length $query and length $grep)
      {
         $patt = "-q'$query'";
         $mode = '-e' if defined $exact;
      }

      @results = `$find | fzf $mode $patt $fzf_opts --preview "$preview"`;
   }
   else
   {
      my $find = "fd -tf $hidden -E.git -E.svn -E.hg --ignore-file ~/.gitignore";
      my $preview = "if file --mime {} | grep -q binary; then echo 'No preview available' 1>&2; else cat {}; fi";

      # search help files matching topic
      if (length $query)
      {
         # fd without query matches anything, thus fzf will filter on whole paths,
         # this is why with query (--exact), -p is needed so query can filter on whole paths too
         my $mode = defined $exact ? "-pF '$query'" : '';

         # -q isn't required with 'exact', it's supplied to enable highlighting
         @results = `$find $mode | fzf -q'$query' $fzf_opts --preview "$preview"`;
      }
      # search trough all help files
      else
      {
         # fuzzy (default) or exact?
         my $mode = defined $exact ? '-e' : '';

         @results = `$find | fzf $mode $fzf_opts --preview "$preview"`;
      }
   }

   chomp @results;
   exit 1 unless grep /\S/, @results; # canceled with Esc or ^C

   $query = $results[0];
   my (undef, $key, $file) = @results;

   # no results: trim any fzf extended search mode characters
   unless ($file)
   {
      $query =~ s/^'//;
      $query =~ tr/^\\$//d;
   }

   return ($query, $key, $file);
}

sub Open
{
   my ($query, $key, $file) = @_;

   # open with your EDITOR
   if ($key or $view)
   {
      my $EDITOR = $ENV{EDITOR} // 'vi';

      # open with nvim (send to running instance)?
      if ($grep and length $query and $EDITOR =~ /vim/i)
      {
         exec $EDITOR, $file, '-c', "0/$query", '-c', 'noh|norm zz<cr>';
      } else {
         exec $EDITOR, $file;
      }
   }

   my (undef, undef, $ext) = fileparse ($file, qr/\.[^.]+$/);

   my $open = $^O eq 'darwin' ? 'open' : 'xdg-open';

   # binary files, is -x test needed?
   if (not -x $file and -B _ || $ext =~ /\.pdf$/i)
   {
      exec $open, $file; # open with adequate app
   }

   # personal help files
   if (-f "$ENV{REPOS_BASE}/help/$file")
   {
      if ($ext =~ /\.md$/i)
      {
         # open markdown docs in the browser
         exec $open, "https://github.com/kurkale6ka/help/blob/master/$file";
      }
      elsif ($file eq 'printf.pl')
      {
         say CYAN, ($dir ne '.' ? "$dir/" : ''), $file, RESET;
         do "./$file";
         exit;
      }
   }

   # display path of file being viewed
   say CYAN, ($dir ne '.' ? "$dir/" : ''), $file, RESET;

   # cat
   if ($ext =~ /\.(?!te?xt).+$/i)
   {
      exec qw/bat --style snip --italic-text always --theme zenburn -mconf:ini/, $file;
   } else {
      exec 'cat', $file;
   }
}

sub Grep
{
   my ($query, $grep) = @_;
   my ($query_bak, $key, $file);

   do {
      $query_bak = $query;
      ($query, $key, $file) = fzf $query, $grep;
   }
   until ($file);

   if ($only) # grep -o
   {
      exec qw/rg -FS/, $query_bak, $file;
   } else {
      Open $query_bak, $key, $file;
   }
}

# Main

# multiple args, split @ARGV with -e?
# rg -Sl patt1 | ... | xargs rg -S pattn
# also for fd ... $1
my $query = shift;
my $file;

# Search files matching topic

# grep -l | fzf
Grep ($query // '', $grep) if defined $grep;

# find | fzf
my @results = defined $query ? fzf $query : fzf;

($query, undef, $file) = @results;

# filename match
if ($file)
{
   Open @results;
} else {
   # list files with occurrences of topic
   Grep ($query, $grep // '');
}

File: ex.py
#! /usr/bin/env python3

"""Fuzzy File Explorer
"""

import argparse
from dataclasses import dataclass
from pathlib import Path
from os import environ as env, execlp, chdir, PathLike
from subprocess import Popen, PIPE
import webbrowser as browser
from shutil import which

# TODO:
# --exact?
# --grep --only
# recursive grep...

parser = argparse.ArgumentParser()
parser.add_argument(
    "-s",
    "--source-dir",
    type=str,
    default=".",
    nargs="?",
    help="define source directory",
)
parser.add_argument(
    "--header",
    action=argparse.BooleanOptionalAction,
    default=True,
    help="show file path",
)
parser.add_argument(
    "--hidden",
    action=argparse.BooleanOptionalAction,
    default=True,
    help="show dotfiles",
)
grp_view = parser.add_mutually_exclusive_group()
grp_view.add_argument(
    "-b", "--view-in-browser", action="store_true", help="view in browser"
)
grp_view.add_argument(
    "-v",
    "--view-in-editor",
    action="store_true",
    help="view in $EDITOR, use alt-v from within fzf",
)
parser.add_argument("-g", "--grep", type=str, help="list files with matches")
parser.add_argument("query", type=str, nargs="?", help="fzf query")
args = parser.parse_args()


# TODO: @contextmanager
class cd:
    """Context manager for changing the current working directory"""

    def __init__(self, newPath):
        self.newPath = Path(newPath).expanduser()

    def __enter__(self):
        self.savedPath = Path.cwd()
        chdir(self.newPath)

    def __exit__(self, etype, value, traceback):  # pyright: ignore reportUnusedVariable
        chdir(self.savedPath)


class Command:
    def __init__(self, cmd: list[str] = []):
        self._cmd = cmd

    @property
    def cmd(self) -> list[str]:
        return self._cmd

    @cmd.setter
    def cmd(self, value: list[str]):
        self._cmd = value


class Search(Command):
    # TODO: both should find the same amount files, but this is not the case when tested in ~
    fd = [
        "fd",
        "--strip-cwd-prefix",
        "-tf",
        "-p",
        "--ignore-file",  # needed since I want ignored files to be also ignored in non .git folders
        f"{env['XDG_CONFIG_HOME']}/git/ignore",
    ]
    rg = [  # TODO: --binary? test with .pdfs
        "rg",
        "-S",
        "-l",
        "--ignore-file",
        f"{env['XDG_CONFIG_HOME']}/git/ignore",
    ]

    def __init__(self, hidden=True, pattern=None):
        if hidden:
            Search.fd.append("--hidden")
            Search.rg.append("--hidden")
        self._pattern = pattern
        if self._pattern:
            Search.rg.append(self._pattern)
            super().__init__(cmd=Search.rg)
        else:
            super().__init__(cmd=Search.fd)

    @property
    def pattern(self):
        return self._pattern

    @property
    def is_grep(self) -> bool:
        return True if self._pattern else False


class Filter(Command):
    fzf = ["fzf", "-0", "-1", "--cycle", "--print-query", "--expect=alt-v"]

    def __init__(self, query=None, pattern=None):
        self._query = query
        if self._query:
            Filter.fzf.extend(("-q", self._query))
        if pattern:
            # TODO: show whole file with lines highlighted?
            Filter.fzf.extend(("--preview", f"rg -FS --color=always '{pattern}' {{}}"))
        else:
            Filter.fzf.extend(("--preview", f"bat --color always {{}}"))
        super().__init__(cmd=Filter.fzf)

    @property
    def query(self):
        return self._query


@dataclass
class FilterResults:
    search_cmd: Search
    filter_query: str | None = None
    pressed_keys: str | None = None
    document: str | None | PathLike = None

    def __post_init__(self):
        if self.filter_query == "":
            self.filter_query = None
        if self.pressed_keys == "":
            self.pressed_keys = None
        if self.document == "":
            self.document = None


class Viewer(Command):
    def __init__(self, prog: str = "cat", header: bool = True):
        self._prog = prog
        self._header = header

    # TODO: be able to: viewer = 'editor' which would assign to prog
    @property
    def prog(self):
        return self._prog

    @prog.setter
    def prog(self, value: str):
        self._prog = value

    @property
    def header(self):
        return self._header

    @header.setter
    def header(self, value: bool):
        self._header = value

    def show(self):
        execlp(*self._cmd)

    def __eq__(self, other: str) -> bool:
        return self._prog == other


class Documents:
    def __init__(self, src: str | PathLike = ".", viewer=Viewer()):
        self._src = src
        self._viewer = viewer

    def search(self, command: Search, filter: Filter):
        with cd(self._src):
            p1 = Popen(command.cmd, stdout=PIPE, text=True)
            p2 = Popen(filter.cmd, stdin=p1.stdout, stdout=PIPE, text=True)
            p1.stdout.close()  # pyright: ignore reportOptionalMemberAccess
            results = p2.communicate()[0].rstrip()  # get stdout

            # fzf returns this string:
            #     --print-query\n
            #     --expect\n
            #     document
            results = results.split("\n")
            data = FilterResults(command, *results)

            if data.document:
                self._open(data)
            elif data.filter_query:
                # trim any fzf extended search mode characters
                data.filter_query = data.filter_query.lstrip("^'")
                data.filter_query = data.filter_query.rstrip("$")
                data.filter_query = data.filter_query.replace("\\", "")

                if not command.is_grep:
                    self.search(
                        Search(pattern=data.filter_query),
                        Filter(pattern=data.filter_query),
                    )
            else:
                exit(1)

    def _open(self, data):
        if self._viewer == "editor" or data.pressed_keys:
            editor = env.get("EDITOR", "vi")
            self._viewer.cmd = [editor, editor, data.document]

            # open with nvim (send to running instance)?
            if data.search_cmd.is_grep and "vim" in editor:
                self._viewer.cmd = [
                    editor,
                    editor,
                    "-c",
                    f"0/{data.search_cmd.pattern}",
                    "-c",
                    "noh|norm zz<cr>",
                    data.document,
                ]
            self._viewer.show()

        extension = Path(data.document).suffix

        # Personal help files
        if Path(f"{env['REPOS_BASE']}/github/help/{data.document}").is_file():
            if self._viewer == "browser" and extension in (".adoc", ".md", ".rst"):
                browser.open(
                    f"https://github.com/kurkale6ka/help/blob/master/{data.document}"
                )
                exit()

            if Path(data.document).name == "printf.pl":
                execlp("perl", "perl", data.document)

        if self._viewer == "cat":
            if self._viewer.header:
                print("File:", Path(data.document).resolve())

            if which("bat") is not None and extension not in (".txt", ".text"):
                self._viewer.cmd = ["bat", "bat", data.document]
            else:
                self._viewer.cmd = ["cat", "cat", data.document]

        self._viewer.show()


if __name__ == "__main__":
    viewer = Viewer(header=args.header)

    if args.view_in_browser:
        viewer.prog = "browser"

    if args.view_in_editor:
        viewer.prog = "editor"

    search_params = dict(hidden=args.hidden)
    filter_params = dict()

    if args.grep:
        search_params["pattern"] = args.grep
        filter_params["pattern"] = args.grep

    if args.query:
        filter_params["query"] = args.query

    docs = Documents(src=args.source_dir, viewer=viewer)
    docs.search(Search(**search_params), Filter(**filter_params))
